### ISSUE TITLE
Options for modules/sway/window: empty, only_floating, only_tiling

### ISSUE THREAD

**Issue by dmyTRUEk** [Alexays/Waybar#1399]
Currently there are only 3 options in css for window: `empty`, `solo`, `<app_id>`.

In 0.9.8 `waybar::modules::sway::leafNodesInWorkspace` returned only the amount of tiling windows, but in 0.9.9 (changed in [this commit](https://github.com/Alexays/Waybar/commit/99723845976fb577a057b86e36b502c966bf254f#diff-8b4bd9b29db6e716716525789b945720af9e21335d52922fde1e4fc0ca422f3a)) it was changed to return the amount of tiling and floating windows.

On 0.9.8 i had this in my `style.css`:
```
window#waybar.empty {
    background-color: transparent;
}
```
and had transparent bar when there were no windows at all, or when there were only floating windows.

But now (on 0.9.9) it's impossible to configure bar color for the case when i don't have tiling windows, but have floating windows :(

I think css could have some more options (smt like this): `only_floating`, `only_tiling`, maybe even `solo_floating`, `solo_tiling`, which will be quite usable.

**Comment by RobertMueller2** [comment]
Sorry, that was my fault, I didn't think there was a use case in ignoring floating windows. ;)

There are a lot of different possibilities:

empty
0 tiled and 1 floating
0 tiled and n floating
1 tiled and 0 floating
1 tiled and 1 floating
1 tiled and n floating
n tiled and 0 floating
n tiled and 1 floating
n tiled and n floating

Personally, I think when adjusting this, they should all somehow be covered. Either the code gets somewhat uglier, or we just add classes along the lines of 0-tiled or 1-tiled, or n-tiled as well as 0-float or 1-float or n-float, and then the user has to style the different combinations to their liking. Either way it inevitably breaks existing styles.

Even though I'm not terribly comfortable with or experienced in C++, I'm not totally unwilling to try and help here. (No promises on time though.)

But it would be good if @Alexays could point out a direction.

**Comment by RobertMueller2** [comment]
Just another thought. Alternatively, there could be a config option to decide between old and new behaviour? Has the advantage of not breaking existing styles.

**Comment by dmyTRUEk** [comment]
In my opinion, the best solution is to add new options, without changing behavior of an old options (comparing to 0.9.9 or 0.9.8?).

**Comment by deuill** [comment]
I've got a similar use-case that I think might be solved by a more general approach here -- I use smart gaps in Sway, which means that a solo window or tabbed window container will take up the entire screen space, whereas tiled windows (or floating windows) only will leave a gap in screen edges.

I've currently got Waybar have a transparent background when there's no windows, or a solid background (which blends with Sway decorations) in any other case -- ideally I'd be able to have something like this, though:

```css
#waybar {
    background-color: #282828;
    color: #d8d8d8;
}

#waybar.empty, #waybar.tiled, #waybar.floating {
    background-color: transparent;
}
```

Essentially, assign a class to the `#waybar` container based on what the layout is for the current workspace, with the following special rules:

- No normal or floating nodes means an `empty` class.
- No normal nodes, but any amount of floating nodes means a `floating` class.
- A single normal node with any amount of floating nodes means a `solo` class.
- Multiple normal nodes in `splitv` or `splith` layouts, with any amount of floating nodes, means a `tiled` class.
- Multiple normal nodes in a `tabbed` layout, with any amount of floating nodes, means a `tabbed` class.

Does this sound about right? It does seem to map to existing use for the `empty` and `solo` classes, and may solve the original issue as well (only the previous behaviour does indeed change and needs to be captured via the `floating` class).

**Comment by Iss-in** [comment]
yes, currently whenver a window is ficused which is floating by default, waybar is treating it as empty workspace. this should not be the desired behaviour

**Comment by RobertMueller2** [comment]
> yes, currently whenver a window is ficused which is floating by default, waybar is treating it as empty workspace. this should not be the desired behaviour

@Iss-in even with 0.99 or current git?

**Comment by Iss-in** [comment]
yes, with both release(0.9.9-1) and -git version from AUR

**Comment by RobertMueller2** [comment]
> I've got a similar use-case that I think might be solved by a more general approach here -- I use smart gaps in Sway, which means that a solo window or tabbed window container will take up the entire screen space, whereas tiled windows (or floating windows) only will leave a gap in screen edges.
> 
> I've currently got Waybar have a transparent background when there's no windows, or a solid background (which blends with Sway decorations) in any other case -- ideally I'd be able to have something like this, though:
> 
> ```css
> #waybar {
>     background-color: #282828;
>     color: #d8d8d8;
> }
> 
> #waybar.empty, #waybar.tiled, #waybar.floating {
>     background-color: transparent;
> }
> ```
> 
> Essentially, assign a class to the `#waybar` container based on what the layout is for the current workspace, with the following special rules:
> 
>     * No normal or floating nodes means an `empty` class.
> 
>     * No normal nodes, but any amount of floating nodes means a `floating` class.
> 
>     * A single normal node with any amount of floating nodes means a `solo` class.
> 
>     * Multiple normal nodes in `splitv` or `splith` layouts, with any amount of floating nodes, means a `tiled` class.
> 
>     * Multiple normal nodes in a `tabbed` layout, with any amount of floating nodes, means a `tabbed` class.
> 
> 
> Does this sound about right? It does seem to map to existing use for the `empty` and `solo` classes, and may solve the original issue as well (only the previous behaviour does indeed change and needs to be captured via the `floating` class).


I think if we add stacked, it covers all use cases. Personally I'd be a bit concerned though about how to do it right, without ending up with code akin to a bowl of spaghetti. ;) Counting nodes and/or floating nodes is one thing, but to decide on the layout, the trees can get fairly complex, even somewhat unexpected, depending on what hotkeys you happened to hit. Take this tree for example, I have no idea what key combos I entered on workspace 12, but on workspace 15 I certainly don't remember setting the workspace to tabbed before opening floating Steam. Tabbed can only be read from a con (I think, who knows?), but splitv/splith I might have to consider from the workspace and the node count immediately underneath.

![tree](https://user-images.githubusercontent.com/29629287/152112913-a6c2cb25-a1ef-4494-bd2a-13836f05b53f.png)

Sure, you can probably always decide /something/, but it might be totally different from what the user expects. I think There would have to be a logging opportunity that makes it somewhat transparent why a certain class was set or added.

**Comment by RobertMueller2** [comment]
I think if we add stacked, it covers all use cases. Personally I'd be a bit concerned though about how to do it right, without ending up with code akin to a bowl of spaghetti. ;) Counting nodes and/or floating nodes is one thing, but to decide on the layout, the trees can get fairly complex, even somewhat unexpected, depending on what hotkeys you happened to hit. Take this tree for example, I have no idea what key combos I entered on workspace 12, but on workspace 15 I certainly don't remember setting the workspace to tabbed before opening floating Steam. Tabbed can only be read from a con (I think, who knows?), but splitv/splith I might have to consider from the workspace and the node count immediately underneath.

![tree](https://user-images.githubusercontent.com/29629287/152114526-adb73fca-9047-4306-8f28-a47044fb02f0.png)

Sure, you can always decide /something/, but it might be different from what the user expects. There'd have to be a logging opportunity to make it transparent why the module decided to set or add a certain class.

**Comment by deuill** [comment]
I was reading through the output of `swaymsg -t get_tree` while I was writing the above, and indeed I was considering the `layout` parameter on the workspace only (like the final parameter that appears in the tree for `type=workspace` and *regardless* of what the tree underneath looks like.

Noted that it wouldn't matter what the workspace layout is if all you have is a single normal node (which means a `solo` class) or any non-zero amount of floating nodes (which means a `floating` class -- the layout only comes into play once you have more than one normal nodes in the workspace (alongside any amount of floating nodes). My assumption here is that the workspace represents the top-most container relevant to Waybar, and that Waybar appears at most once per workspace (but that you may have multiple workspaces visible, i.e. by having multiple screens; not sure if each instance gets its own class there).

**Comment by RobertMueller2** [comment]
Ok, that sounds fair enough. Perhaps add an FAQ entry to check get_tree for workspace layout in case of unexpected styling, and that way it might not even be too bad. ;)

Regarding multiple screens, there seems to be an undocumented (well at least it's not in the manpage) boolean config setting "all-outputs" which controls this.

If false, nothing is displayed unless the focused node's output is the same of the waybar. If true, the focused node is displayed an all waybars.

**Comment by WuerfelDev** [comment]
I haven't read through the code, but I noticed a related bug. On my multi-monitor setup waybar on the monitor that is not active does not have the class `solo` but always `empty`. Running waybar 0.9.9

For example I have one window on each monitor. I'd expect that both waybars show that there is a `solo` window on their monitor. But currently the active one shows as `solo` and the other one shows as `empty`. That switches once the mouse enters the other monitor

**Comment by RobertMueller2** [comment]
@WuerfelDev I don't think this is a new issue, if I had to speculate I'd say it's because the module is starting from the focused node, and on your other not focused screen there currently is no focused node. 

I'm not opposed to styling each screen's waybar according to the current workspace on a screen, but I think it would be a bigger change.

But what would the text be on the not focused screen? After all, "The window module displays the title of the currently focused window in Sway". Personally, I'd rather have it empty on the off-screen. Perhaps there could be options to control all this.

**Comment by RobertMueller2** [comment]
@WuerfelDev I have implemented something in PR #1419. When setting config option "offscreen-text" to anything non-empty, the given text will be displayed on the currently not focused screen along with styling for the current window situation on that screen. This is ignored when "all-outputs" is enabled. I hope that's what you wanted. ;)

**Comment by harogaston** [comment]
Sorry, related but actually looking for help. I have this in my css:

```css
window#waybar.solo {
  padding: 0px;
  margin: 0px;
  border: 0px;
  background-color: #FF0000;
}

window#waybar.empty {
	background-color: #00FF00;
}
```

but waybar does not change in any way. It works only if I remove the `.solo` or `.empty` but those are the ones I am interested in.
Do you know what may be happening? (I have "sway/mode" in my modules)
Thank you.

**Comment by RobertMueller2** [comment]
@harogaston sway/mode does not matter for these styles. They are applied by the sway/window module. Do you have that one in your config?

The css itself should apply to the whole waybar. If that's what you want it looks correct to me, but it needs sway/window.

**Comment by harogaston** [comment]
Oh thank you! I had read the documentation but it seems I didn't quite understood that part. Thank you I got the expected behavior now.

Cheers!

**Comment by Alexays** [comment]
This is now implemented on master. The sway/window module sets these classes on the bar based on the current workspace layout:

- `empty` — no tiled and no floating windows
- `floating` — no tiled windows but one or more floating windows (this is the "only floating" case you asked for)
- `solo` — a single tiled window
- `tabbed` / `stacked` / `tiled` — multiple windows, by layout

So the exact case you described — no tiling windows but floating windows present — can now be styled via `window#waybar.floating`, and a truly empty workspace via `window#waybar.empty`. See src/modules/sway/window.cpp.

Closing as resolved, but please reopen if a combination you need still isn't covered. Thanks for the detailed report! 🙏
